### ISSUE TITLE
`torch.func.functional_call` does not work with `__torch_function__ ` Tensor-like objects

### ISSUE THREAD

**Issue by samuela** [pytorch/pytorch#110249]
### 🐛 Describe the bug

The [documentation](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-with-a-tensor-like-type) indicates that `torch` can be extended with `torch.Tensor`-like objects via the `__torch_function__` mechanism for dynamic dispatch. However, `torch.func.functional_call` does not accept such objects as inputs. For example,
```py
# begin code copied from the documentation: https://pytorch.org/docs/stable/notes/extending.html#extending-torch-with-a-tensor-like-type
HANDLED_FUNCTIONS = {}
class ScalarTensor(object):
  def __init__(self, N, value):
    self._N = N
    self._value = value

  def __repr__(self):
    return "ScalarTensor(N={}, value={})".format(self._N, self._value)

  def tensor(self):
    return self._value * torch.eye(self._N)

  @classmethod
  def __torch_function__(cls, func, types, args=(), kwargs=None):
    if kwargs is None:
      kwargs = {}
    if func not in HANDLED_FUNCTIONS or not all(
        issubclass(t, (torch.Tensor, ScalarTensor))
        for t in types
    ):
      return NotImplemented
    return HANDLED_FUNCTIONS[func](*args, **kwargs)

import functools
def implements(torch_function):
  """Register a torch function override for ScalarTensor"""
  def decorator(func):
    functools.update_wrapper(func, torch_function)
    HANDLED_FUNCTIONS[torch_function] = func
    return func
  return decorator

@implements(torch.mean)
def mean(input):
  return float(input._value) / input._N
# end code copied from documentation

model = torch.nn.Linear(3, 3, bias=False)
input = torch.randn(5, 3)
torch.func.functional_call(model, {"weight": ScalarTensor(3, 0.1)}, (input,))
```
produces an error
```
/usr/local/lib/python3.10/dist-packages/torch/nn/utils/_named_member_accessor.py in swap_tensor(module, name, tensor, allow_missing)
     40         and tensor is not None
     41     ):
---> 42         raise TypeError(f"{tensor} is not an instance of torch.Tensor")
     43     if "." in name:
     44         raise KeyError('tensor name can\'t contain "."')
```

### Versions

Collecting environment information...
PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.4
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.120+-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          2
On-line CPU(s) list:             0,1
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                      6
Model:                           79
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
Stepping:                        0
BogoMIPS:                        4399.99
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB (1 instance)
L1i cache:                       32 KiB (1 instance)
L2 cache:                        256 KiB (1 instance)
L3 cache:                        55 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.1+cu118
[pip3] torchaudio==2.0.2+cu118
[pip3] torchdata==0.6.1
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.15.2
[pip3] torchvision==0.15.2+cu118
[pip3] triton==2.0.0
[conda] Could not collect

**Comment by samuela** [comment]
Yeah, I think that should do the trick although admittedly I'm not confident in the internals here.

**Comment by ezyang** [comment]
literally just remove the line mentioned in the backtrace
